src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs

```diff
@@ -293,114 +344,269 @@ public bool TryGetValue(string key, out object value)
                 throw new ArgumentNullException(nameof(key));
             }
 
-            if (InnerDictionary != null)
+            return _storage.TryGetValue(key, out value);
+        }
+
+        private void Upgrade()
+        {
+            _storage.Upgrade(ref _storage);
+        }
+
+        public struct Enumerator : IEnumerator<KeyValuePair<string, object>>
+        {
+            private readonly Storage _storage;
+            private int _index;
+
+            public Enumerator(RouteValueDictionary dictionary)
+            {
+                if (dictionary == null)
+                {
+                    throw new ArgumentNullException();
+                }
+
+                _storage = dictionary._storage;
+
+                Current = default(KeyValuePair<string, object>);
+                _index = -1;
+            }
+
+            public KeyValuePair<string, object> Current { get; private set; }
+
+            object IEnumerator.Current => Current;
+
+            public void Dispose()
```

Comment:
MSDN will [explain it better than I can](https://msdn.microsoft.com/en-us/library/system.idisposable.dispose(v=vs.110).aspx)